### PR TITLE
[Feature]Support Relational Knowledge Distillation

### PR DESCRIPTION
**Pull request by HIT-cwh**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/models/losses/rkd.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        Args:
+            preds_S (torch.Tensor): The student model prediction with
+                shape (N, C, H, W) or shape (N, C).
+            preds_T (torch.Tensor): The teacher model prediction with
+                shape (N, C, H, W) or shape (N, C).
+        Return:
+            torch.Tensor: The calculated loss value.
+        """
+        preds_S = preds_S.view(preds_S.shape[0], -1)
+        preds_T = preds_T.view(preds_T.shape[0], -1)
```
Comment (humu789) [inline mmrazor/models/losses/rkd.py:94]: `preds_S = preds_S.view(preds_S.shape[0], -1)`
 `preds_T = preds_T.view(preds_T.shape[0], -1)`
they seem redundant

File: mmrazor/models/losses/rkd.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        if self.l2_norm:
+            preds_S = F.normalize(preds_S, p=2, dim=-1)
+            preds_T = F.normalize(preds_T, p=2, dim=-1)
```
Comment (humu789) [inline mmrazor/models/losses/rkd.py:97]: arg `dim` is more of 1 than -1

File: configs/distill/rkd/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# WSLD
```
Comment (humu789) [inline configs/distill/rkd/README.md:1]: RKD

File: configs/distill/rkd/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+![pipeline](/docs/en/imgs/model_zoo/rkd/pipeline.png)
+
+## Results and models
+### Classification
+|Location|Dataset|Teacher|Student|Acc|Acc(T)|Acc(S)|Config | Download |
+:--------:|:---------:|:---------:|:---------:|:---------:|:---------:|:---------:|:------:|:---------|
+| neck     |ImageNet|[resnet34](https://github.com/open-mmlab/mmclassification/blob/master/configs/resnet/resnet34_8xb32_in1k.py)|[resnet18](https://github.com/open-mmlab/mmclassification/blob/master/configs/resnet/resnet18_8xb32_in1k.py)| 70.23 |    73.62 |    69.90  |[config](./rkd_neck_resnet34_resnet18_8xb32_in1k.py)|[teacher](https://download.openmmlab.com/mmclassification/v0/resnet/resnet34_b16x8_cifar10_20210528-a8aa36a6.pth) &#124;[model]() &#124; [log]()|
+| neck     |Cifar100|[resnet50](https://github.com/open-mmlab/mmclassification/blob/master/configs/resnet/resnet50_8xb16_cifar100.py)|[vgg11bn]()| 72.76 |    79.90 |    71.26  |[config](./rkd_neck_resnet50_vgg11bn_8xb16_cifar100.py)|[teacher](https://download.openmmlab.com/mmclassification/v0/resnet/resnet50_b16x8_cifar100_20210528-67b58a1b.pth) &#124;[model]() &#124; [log]()|
```
Comment (humu789) [inline configs/distill/rkd/README.md:30]: why verify RKD with these model rather than some models in papers

File: mmrazor/models/losses/rkd.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    Args:
+        loss_weight_d (float): Weight of distance-wise distillation loss.
+            Defaults to 25.0.
+        loss_weight_a (float): Weight of angle-wise distillation loss.
+            Defaults to 50.0.
+        l2_norm (bool): Whether to normalize the model predictions before
+            calculating the loss. Defaults to True.
+    """
+
+    def __init__(self, loss_weight_d=25.0, loss_weight_a=50.0, l2_norm=True):
```
Comment (humu789) [inline mmrazor/models/losses/rkd.py:23]: why is `loss_weight_d` : `loss_weight_a` 25 : 50 rather than 1 : 2 ?

**Review by humu789 — CHANGES_REQUESTED** [review]

File: mmrazor/models/losses/rkd.py (hunk 1 above)
Comment (HIT-cwh) [reply mmrazor/models/losses/rkd.py:94]: The shape of prediction may be (N, C, H, W) ?

File: mmrazor/models/losses/rkd.py (hunk 2 above)
Comment (HIT-cwh) [reply mmrazor/models/losses/rkd.py:97]: Fix, thx!

File: configs/distill/rkd/README.md (hunk 3 above)
Comment (HIT-cwh) [reply configs/distill/rkd/README.md:1]: Fixed, thx!

File: configs/distill/rkd/README.md (hunk 4 above)
Comment (HIT-cwh) [reply configs/distill/rkd/README.md:30]: R50-vgg11bn(cifar100) is the settings in the paper. Experiments on TinyImageNet are not conducted as mmcls does not support this dataset.  Add R34-R18 (ImageNet) to verify its effectiveness on a larger dataset and it can be seen as an algorithm in Distillation Benchmark.

File: mmrazor/models/losses/rkd.py (hunk 5 above)
Comment (HIT-cwh) [reply mmrazor/models/losses/rkd.py:23]: The paper set ``loss_weight_d`` = 25 and ``loss_weight_a`` = 50 by default.

Comment (humu789) [reply mmrazor/models/losses/rkd.py:23]: OK, you are right, it seems to be 1:2 in official github

File: configs/distill/rkd/README.md (hunk 4 above)
Comment (humu789) [reply configs/distill/rkd/README.md:30]: Got it.

File: mmrazor/models/losses/rkd.py (hunk 1 above)
Comment (humu789) [reply mmrazor/models/losses/rkd.py:94]: It seems to lack a `nn.Linear` layer below these lines, otherwise can not set arg `embedding_size`

File: configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    neck=dict(type='GlobalAveragePooling'),
+    head=dict(
+        type='LinearClsHead',
+        num_classes=1000,
+        in_channels=512,
+        loss=dict(type='CrossEntropyLoss', loss_weight=1.0),
+        topk=(1, 5),
+    ))
+
+# teacher settings
```
Comment (pppppM) [inline configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py:25]:
```suggestion
# teacher settings
teacher_ckpt=xxxxx
```

File: configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+teacher = dict(
+    type='mmcls.ImageClassifier',
```
Comment (pppppM) [inline configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py:27]:
```suggestion
    type='mmcls.ImageClassifier',
    init_cfg=dict(type='Pretrained', checkpoint=teacher_ckpt),
```
Many users confuse the usage of `init_cfg`, and it is better to write directly in config

File: mmrazor/models/losses/rkd.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        super(RelationalKD, self).__init__()
+        self.loss_weight_d = loss_weight_d
+        self.loss_weight_a = loss_weight_a
+        self.l2_norm = l2_norm
```
Comment (pppppM) [inline mmrazor/models/losses/rkd.py:27]: `l2_norm` -> `with_l2_norm`

File: mmrazor/models/losses/rkd.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+        Args:
+            pred (torch.Tensor): The prediction of the teacher or student with
+                shape (N, C).
+            squared (bool): Whether to calculate the squared Euclidean
+                distance. Defaults to False.
+            eps (float): The minimum Euclidean distance between the two
+                examples. Defaults to 1e-12.
+        """
+        pred_square = pred.pow(2).sum(dim=-1)
```
Comment (pppppM) [inline mmrazor/models/losses/rkd.py:41]: It is better to add some doc-string to describe the process of shape change

**Review by pppppM — CHANGES_REQUESTED** [review]

File: configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py (hunk 7 above)
Comment (HIT-cwh) [reply configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py:27]: It's a good point, but users may need to modify the config file themselves to run our code.

File: mmrazor/models/losses/rkd.py (hunk 8 above)
Comment (HIT-cwh) [reply mmrazor/models/losses/rkd.py:27]: Fixed, thx!

File: mmrazor/models/losses/rkd.py (hunk 9 above)
Comment (HIT-cwh) [reply mmrazor/models/losses/rkd.py:41]: Fixed, thx!

File: configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py (hunk 7 above)
Comment (pppppM) [reply configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py:27]: It's just default ckpt, the user also can modify it by `cfg-options`

Comment (HIT-cwh) [reply configs/distill/rkd/rkd_neck_resnet34_resnet18_8xb32_in1k.py:27]: Good point, fixed.

File: configs/distill/rkd/rkd_neck_resnet50_vgg11bn_8xb16_cifar100.py [hunk 10]
```diff
@@ -18,8 +18,13 @@
     ))
 
 # teacher settings
+
+# FIXME: replace it with your own path
+teacher_ckpt = 'path/to/your/checkpoint.pth'
```
Comment (pppppM) [inline configs/distill/rkd/rkd_neck_resnet50_vgg11bn_8xb16_cifar100.py:23]: There should be mmcls HTTPS link.

**Review by pppppM — CHANGES_REQUESTED** [review]

**Comment by pppppM** [comment]
The `RelationalKD` should split to `RelationalDistLoss` and `RelationalAngleLoss`, and rename the `rkd.py`

**Review by pppppM — APPROVED** [review]

**Review by humu789 — APPROVED** [review]
